### PR TITLE
[Transform] Improve robustness of checkpointing

### LINKED ISSUE #75780: [Transform] Checkpointing should not fail due to replica unavailability
When transform creates a checkpoint it uses the [index stats API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-stats.html) to get the global checkpoints for every shard. 

By design (of index stats) this call queries _every_ shard, including replica shards. Internally this information is de-duplicated and `max(gcp)` is taken.

Fleet implemented very similar functionality with the [_global checkpoints API_](https://www.elastic.co/guide/en/elasticsearch/reference/7.x//get-global-checkpoints.html). The implementation for it does **not** query replica shards and only collects the information that is required.

Transform can benefit from a similar implementation, if possible by re-using code/functionality.

This will make transform less likely to fail and reduce the amount of network calls for checkpointing by 50 or more percent.

### PR DESCRIPTION
**Pull request by hendrikmuhs**
rewrites checkpointing as internal actions, reducing several sub-calls to
only 1 per data node that has at least 1 primary shard of the indexes of
interest.

Notes

Robustness: The current checkpointing sends a request to every shard - primary and replica - and collects the results. If 1 request fails, even for a replica, checkpointing fails. See #75780 for details.

Performance: The current checkpointing is wasteful, it uses get index and get index stats which results in a lot more calls and executes a lot more code which produces results we are not interested in.

### Number of node<->node messages:

old: `1 + shards * indices * (replicas + 1)`
new: `data_nodes`
(super precise: all data nodes with at least 1 primary shard of the requested indices)

e.g.

|shards|replicas|indices|data_nodes|old|new|
| ------: | --------: | -------: | ------: | ---: | ----: |
|1|1|1|1|4|1|
|5|1|1|1|11|1|
|5|1|5|3|51|3|
|5|1|20|8|201|8|
|5|1|100|10|1001|10|

Fixes #75780

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            if (obj == this) {
+                return true;
+            }
+            if (obj == null || obj.getClass() != getClass()) {
+                return false;
+            }
+            Response that = (Response) obj;
+
+            return this.checkpoints.size() == that.checkpoints.size()
+                && this.checkpoints.entrySet().stream().allMatch(e -> Arrays.equals(e.getValue(), that.checkpoints.get(e.getKey())));
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:126]: Why can't we use `Objects.equals` here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest {
+
+        private final String[] indices;
+        private final IndicesOptions indicesOptions;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            indices = in.readStringArray();
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:44]: Does this handle `null`?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointActionRequestTests.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+
+public class GetCheckpointActionRequestTests extends AbstractWireSerializingTestCase<Request> {
+
+    @Override
+    protected Request createTestInstance() {
+        return new Request(
+            generateRandomStringArray(10, 10, false, false),
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointActionRequestTests.java:27]: Should we also test `null` here?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        protected AsyncGetCheckpointsFromNodesAction(
+            ClusterState clusterState,
+            Task task,
+            Map<String, Set<ShardId>> nodesAndShards,
+            ActionListener<Response> listener
+        ) {
+            this.task = task;
+            this.listener = listener;
+            this.nodesAndShards = nodesAndShards;
+            nodes = clusterState.nodes();
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:134]:
```suggestion
            this.nodes = clusterState.nodes();
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            localNodeId = clusterService.localNode().getId();
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:135]:
```suggestion
            this.localNodeId = clusterService.localNode().getId();
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            return Collections.emptyMap();
+        }
+
+        final DiscoveryNodes nodes = state.nodes();
+        Map<String, Set<ShardId>> nodesAndShards = new HashMap<>();
+
+        ShardsIterator shardsIt = state.routingTable().allShards(concreteIndices);
+        for (ShardRouting shard : shardsIt) {
+            // only take primary shards, which should be exactly 1, this isn't strictly necessary
+            // and we should consider taking any shard copy, but than we need another way to de-dup
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:104]:
```suggestion
            // and we should consider taking any shard copy, but then we need another way to de-dup
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java (hunk 1 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:126]: `Objects.equals` compares references but not the content:

```
    public void testEquals() {
        int[] arr1 = new int[10]; // initialized with 0 in all cells
        int[] arr2 = new int[10]; // initialized with 0 in all cells
        //assertTrue(Objects.equals(arr1, arr2)); // false
        assertTrue(Arrays.equals(arr1, arr2)); // true
    }
    ```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java (hunk 2 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:44]: yes, it is serialized as `size: VInt, member1, member2, ....`

The `null`  check happens in the [write method](https://github.com/elastic/elasticsearch/pull/80984/files/550845f5bbf119aa713c944d2f0b19af61514766#diff-153331a2672d0cfa2c31713b8ae1cdb6741927492579f425d5d57290acd5844cR89)

Nevertheless, I got you regarding test coverage.

**Comment by hendrikmuhs** [comment]
@elasticmachine update branch

**Comment by hendrikmuhs** [comment]
@elasticmachine update branch

**Review by przemekwitek — APPROVED** [review]
LGTM

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+
+public class TransportGetCheckpointAction extends HandledTransportAction<Request, Response> {
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:46]: We are relying on the "latest" cluster state here. Should this be a master node action? Or is it acceptable to not be a master node action and potentially have "old" state?

EDIT: OK, looking at Fleet `GetGlobalCheckpointsAction` this is apparently cool. looks good.

**Review by benwtrent — COMMENTED** [review]
Just to be paranoid, does this actually work with remote clusters? I guess the transport client is used between clusters?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                            oneNodeAndItsShards.getValue().iterator().next(),
+                            "Node not found for [{}] shards",
+                            oneNodeAndItsShards.getValue().size()
+                        )
+                    );
+                    return;
+                }
+
+                logger.trace("get checkpoints from node {}", node);
+                transportService.sendRequest(
```
Comment (davidkyle) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:188]:
```suggestion
                transportService.sendChildRequest(
```

Sending a child request parents the request to this task

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointNodeAction.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        Map<String, long[]> checkpointsByIndexOfThisNode = new HashMap<>();
+        for (ShardId shardId : shards) {
+            final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+
+            checkpointsByIndexOfThisNode.computeIfAbsent(shardId.getIndexName(), k -> {
+                long[] seqNumbers = new long[indexService.getIndexSettings().getNumberOfShards()];
+                Arrays.fill(seqNumbers, SequenceNumbers.UNASSIGNED_SEQ_NO);
+                return seqNumbers;
+            });
```
Comment (davidkyle) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointNodeAction.java:58]: Can you assign to the array on the return value of `computeIfAbsent` and save a `get`?

```suggestion
            })[shardId.getId()] = indexShard.seqNoStats().getGlobalCheckpoint();
```

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointNodeAction.java (hunk 9 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointNodeAction.java:58]: 1 node might store more than 1 shard. So it is possible that `computeIfAbsent` is not executed for the 2nd shard, but the get is required. I looked into using `compute` or `merge` but both gets complicated.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+
+public class GetCheckpointNodeAction extends ActionType<GetCheckpointNodeAction.Response> {
+
+    public static final GetCheckpointNodeAction INSTANCE = new GetCheckpointNodeAction();
+
+    // note: this is an internal action
+    public static final String NAME = "cluster:admin/transform/checkpoint[s]";
```
[the remark this answers is not in the document]
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java]: :+1: I tried this originally but had problem, probably I did it for both. 

Anyway, I changed the node action to internal and it seems to work.

**Comment by hendrikmuhs** [comment]
> Just to be paranoid, does this actually work with remote clusters? I guess the transport client is used between clusters?

yes, it should work the same way as the previous approach. One challenge are older remotes, this is handled like "pit", which means it falls back to the old style if the exception caught is "action not found".

To be sure, I will manually test this.

**Review by benwtrent — APPROVED** [review]

**Comment by hendrikmuhs** [comment]
changing the node action to `internal` breaks CCS, I am getting an authorization error. 

I will have another look tomorrow, I don't want to give up, this might just be a bug.

**Comment by benwtrent** [comment]
> changing the node action to internal breaks CCS, I am getting an authorization error.

It may be that since its a separate cluster, it has to travel through a path that is no longer "internal". In that case, I think switching back to `admin` is ok. But, it seems to me that it should work as it is all caused by an internal source.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        resolveIndicesAndGetCheckpoint(task, request, listener, state);
+    }
+
+    protected void resolveIndicesAndGetCheckpoint(Task task, Request request, ActionListener<Response> listener, final ClusterState state) {
+        String[] concreteIndices = this.indexNameExpressionResolver.concreteIndexNames(
+            state,
+            request.indicesOptions(),
+            true, // includeDataStreams
+            request.indices()
+        );
```
Comment (ywangd) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportGetCheckpointAction.java:82]: This does not work because `GetCheckpointAction` is a cluster action (`cluster:admin/transform/checkpoint`). Authorization does not expand indices for cluster actions, this means the expansion here does not respect user's index access and could expand to anything that the cluster has. The quick fix is likely changing the action from `cluster:...` to `indices:...`.

**Review by ywangd — CHANGES_REQUESTED** [review]
The action needs to go through index expansion in the Authorization code. Otherwise, it does not respect user's access.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    public static final GetCheckpointAction INSTANCE = new GetCheckpointAction();
+
+    // note: this is an admin action, it must be called with user headers
+    public static final String NAME = "cluster:admin/transform/checkpoint";
+
+    private GetCheckpointAction() {
+        super(NAME, GetCheckpointAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest {
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:38]: This request also needs to implement `IndicesRequest.Replaceable` so the indices are expandable for Authorization.

**Comment by hendrikmuhs** [comment]
Thank you @ywangd for your help. I further investigated and implemented what we discussed. It would be nice if you can take a look again and provide your feedback.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java [hunk 13]
```diff
@@ -90,7 +90,9 @@
         "cluster:admin/data_frame/*",
         "cluster:monitor/data_frame/*",
         "cluster:monitor/transform/*",
-        "cluster:admin/transform/*"
+        "cluster:admin/transform/*",
+        "cluster:internal/transform/*",
+        "indices:internal/transform/*"
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:95]: `indices:internal/transform/checkpoint` is indexAction in that the request needs index name expansion. This means it cannot be part of a cluster privilege. It really needs be granted as an index privilege. I noticed you already add it as part of `view_index_metadata` index privilege. So I suggest we remove it from here.

For `cluster:internal/transform/*`, I don't know how transform works, so maybe this is normal: But having it here means the end-user needs to have both `manage_transform` cluster privilege and `view_index_metadata` index privilege for the actions to work. Is this expected?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    public static final GetCheckpointAction INSTANCE = new GetCheckpointAction();
+
+    // note: this is an admin action, it must be called with user headers
+    public static final String NAME = "indices:internal/transform/checkpoint";
+
+    private GetCheckpointAction() {
+        super(NAME, GetCheckpointAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest.Replaceable {
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:38]: `IndicesRequest.Replaceable` has a default method `boolean allowsRemoteIndices()` that returns `false`, which means the `indices` field cannot contain remote index names of the format `cluster_alias:index`. My reading of the code seems to suggest that all indices here have their cluster_alias prefix stripped (if any). So it should be OK for the class to rely on the default implementation of `allowsRemoteIndices()`. But it would be great if you could confirm on this.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+
+public class GetCheckpointNodeAction extends ActionType<GetCheckpointNodeAction.Response> {
+
+    public static final GetCheckpointNodeAction INSTANCE = new GetCheckpointNodeAction();
+
+    // note: this is an internal action
+    public static final String NAME = "cluster:internal/transform/checkpoint[n]";
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java:31]: I find it weird that this action is a clusterAction but the request still target indices (shards). I really think this should be an indexAction as well in the format of `indices:internal/transform/checkpoint[n]`. An example in the existing code and also similar to the scenario here is the FieldCaps request which is an indicesRequest:
https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesNodeRequest.java#L27

and the action name is `indices:data/read/field_caps[n]`:
https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L52

Renaming it to be `indices:internal/transform/checkpoint[n]` also means it will be covered by the same `view_index_metadata` index privilege. So I suggest we rename the action to `indices:...` and make the associated request implements `IndicesRequest`.

**Review by ywangd — COMMENTED** [review]
Thanks Hendrik. I have some more suggestions/questions on how privilege works. 

Btw, if I understand the code correctly, the two new actions are trying to do something identical or at least similar to [`GetGlobalCheckpointsAction`](https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointsAction.java#L49-L52). This action is in the fleet plugin and not exposed to other plugins. So not easily usable in Transform. The distributed didn't want to expose it more widely when it was first introduced and kinda considered it as beta. But it [does not seem to have any beta label](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/get-global-checkpoints.html). So maybe it's an useful thing to expose and resuse it today. A potential defficiency is that the GetGlobalCheckPoint API issues child request per shard which is less efficient than grouping them per node especially when the cluster size is large.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java (hunk 13 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java:95]: > the end-user needs to have both manage_transform cluster privilege and view_index_metadata index privilege for the actions to work. Is this expected?

Yes, that's expected. The 2 actions are a replacement for the already existing checkpointing. It currently uses the index stats API's which require `view_index_metadata`. Note: the old implementation must be kept in order to provide backwards compatibility(e.g. for CCS).

Having that said, I would not mind if we find a solution that does not require `view_index_metadata`. At least in the long run.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java (hunk 14 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointAction.java:38]: yes, that's correct. CCS is handled before calling this action: the `source` (indices to query) is split per cluster, get checkpoint is called separately for every cluster. This has been implemented like this for the original checkpointing, which as explained above is kept for BWC. It would be quite complex to separate the 2. Once the old style checkpointing got removed somewhen in future, we might revisit.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java (hunk 15 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java:31]: I stumbled upon the same thing. With `IndicesRequest` we carry the index name which is actually not necessary after resolving the shards, that's the reason why I kept it a cluster action. However it probably doesn't matter to send a few bytes more.

**Comment by hendrikmuhs** [comment]
> Btw, if I understand the code correctly, the two new actions are trying to do something identical or at least similar to [`GetGlobalCheckpointsAction`](https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointsAction.java#L49-L52). This action is in the fleet plugin and not exposed to other plugins. So not easily usable in Transform.

We considered this action and we would have taken the code and moved it around to make it usable. However it did not meet our requirements and the changes would have been quite large. This would neither help us nor fleet.

**Comment by ywangd** [comment]
@hendrikmuhs I think there really is just a single issue. So I'll comment here instead of under each comment for easy consumption.

The issue is: Cluster privilege should not be used to authorize indices actions with actual indices. Concretely for this PR, that means `manage_transform` should not be used for the `GetCheckpointNodeAction`. 

I am aware that you gave this action a name begins with `cluster:...`. But I don't think the name should decide whether the action is cluster or indices. In my opinion, it is the opposite, i.e. the nature of the action decides what name it should have. Since `GetCheckpointNodeAction` carries indices that must be authorized for the user, it is an Indices request and we should name it as such. Although these indices are already authorized in the parent request, it is still much safer to let the child requests run through the index authorization process again* as long as they still carry indices. It is a risky path to conditionally skipping it. Also index authorization needs to run (for both parent and child requests) for things like document level security (DLS) to work. Even though this particular use case does not concern DLS, proactively avoiding index authorization is riksy because too much knowledge is assumed by the caller.

In summary, I strongly recommend that we do something similar to `FieldCapabilitiesRequest` and `FieldCapabilitiesNodeRequest` for the privilege and authorization part. I am happy to jump on a sync if it helps.

> We considered this action and we would have taken the code and moved it around to make it usable. However it did not meet our requirements and the changes would have been quite large. This would neither help us nor fleet.

Sorry I wasn't aware that there were already previous discussions. Thanks for the pointer.

\* Index authorization has been improved a lot since 7.15. A child request with all concrete index names should not incur obvious overhead for authorization.

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc

**Review by ywangd — APPROVED** [review]
LGTM

@hendrikmuhs I appreciate your effort on the iterations and reaching out to us for awareness. Thanks a lot!

> It works because of the wildcard: GetCheckpointAction.NAME + "*"

Yes we intentionally use the suffix wildcard pattern so that privileges of child actions are also granted along with the parent action (assuming child actions are named using the suffix pattern).
